components/server/src/server.rs

```diff
@@ -237,6 +237,15 @@ struct TikvServer<ER: RaftEngine> {
     region_info_accessor: RegionInfoAccessor,
     coprocessor_host: Option<CoprocessorHost<RocksEngine>>,
     to_stop: Vec<Box<dyn Stop>>,
+    /// The services should be stopped BEFORE the gRPC service endpoint stopped.
+    /// If your service need to send some message to clients during shuting
+    /// down, add it to here.
+    ///
+    /// Before adding your service here, also make sure your service:
+    /// - Have well-defined behavior of handling requests during shuting down.
+    /// - After shuting down, other services can handle requests properly too.
+    ///   (i.e. No RPC endpoint relies on your service to work)
+    to_early_stop: Vec<Box<dyn Stop>>,
```

Comment:
What do you think about the `to_early_stop`? @BusyJay

Reply:
The problem of making a lifetime hook here is we need to choose a reasonable method to register them. 

Trivially make something like `Server::add_before_shutdown_hook(Box<dyn FnOnce()>)` seems would be enough for us, however it couples unnecessary logic into the server layer (IMO, it should be purely some platform for dependency injecting.).

Perhaps extend the `Stop` trait to something like `Daemon`, and add the corresponding method to `Runnable`:
```rust
trait Daemon {
  fn stop(self: Box<Self>);
  fn before_shutdown(self: Box<Self>);
}
```

Then we can make this trait to manage the lifetime of some trivial services.

Reply:
The idea of calling certain hooks before shutting down LGTM. And adding a different hook is better than blocking in shutdown. Another method is adding a new synchronous mechanism. All services that want to gracefully shutdown should subscribe the event. Once TiKV is about to shutdown, wakes up all those services first and wait for them before actually shutting down grpc server.